apps/system/utils/fscmd.c

```diff
@@ -961,9 +962,116 @@ static int mount_handler(FAR const char *mountpoint, FAR struct statfs *statbuf,
+ * Name: mountpoint_redirection
+ *
+ * Description:
+ *   mountpoint_redirection : Recursively search mount points from the root.
+ *
+ ****************************************************************************/
+static int mountpoint_redirection(const char *cmd, const char *dirpath, mountpoint_handler_t handler, foreach_mountpoint_t mp_handler)
+{
+	int ret = OK;
+	DIR *dirp = opendir(dirpath);		/* Open the directory */
+
+	if (!dirp) {
+		/* Failed to open the directory */
+		FSCMD_OUTPUT("\t TASH %s: no such directory: %s\n", cmd, dirpath);
+
+		return ERROR;
+	}
+
+	/* Read each directory entry */
+	for (;;) {
+		struct dirent *entryp = readdir(dirp);
+		if (!entryp) {
+			/* Finished with this directory */
+			break;
+		}
+
+		/* Call the handler with this directory entry */
+		if (handler(dirpath, entryp, mp_handler) < 0) {
+			/* The handler reported a problem */
+			ret = ERROR;
+			break;
+		}
+	}
+	closedir(dirp);
+
+	return ret;
+}
+
+/****************************************************************************
+ * Name: mountpoint_handler
+ *
+ * Description:
+ *    mountpoint_handler displays a mount information of the given directory
+ *    if it is a mount point.
+ *
+ ****************************************************************************/
+static int mountpoint_handler(FAR const char *dirpath, FAR struct dirent *entryp, foreach_mountpoint_t handler)
+{
+	int ret;
+	struct statfs buf;
+	char *fullpath = get_dirpath(dirpath, entryp->d_name);
+
+	/* Call statfs with the given file path */
+	ret = statfs(fullpath, &buf);
+	if (ret != 0) {
+		FSCMD_OUTPUT(CMD_FAILED, "mountpoint_handler", "");
+		return ERROR;
+	}
+
+	ret = handler(fullpath, &buf, NULL);
+	fscmd_free(fullpath);
+	if (ret != 0) {
+		FSCMD_OUTPUT(CMD_FAILED, "mountpoint_handler", "");
+		return ERROR;
+	}
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: mountpoint_recursive
+ *
+ * Description:
+ *    mountpoint_recursive recursively search all mount points and display
+ *    those using mountpoint_handler.
+ *
+ ****************************************************************************/
+static int mountpoint_recursive(const char *dirpath, struct dirent *entryp, foreach_mountpoint_t handler)
+{
+	int ret = OK;
+
+	if (DIRENT_ISDIRECTORY(entryp->d_type) && !ls_specialdir(entryp->d_name)) {
+		char *newpath;
+		newpath = get_dirpath(dirpath, entryp->d_name);
+
+		/* Traverse the directory */
+		ret = mountpoint_redirection("mountpoint_recursive", newpath, mountpoint_handler, handler);
+		if (ret == OK) {
+			ret = mountpoint_redirection("mountpoint_recursive", newpath, mountpoint_recursive, handler);
+		}
+		fscmd_free(newpath);
+	}
+
+	return ret;
+}
+
+static int print_mountpoint(foreach_mountpoint_t handler, FAR void *arg)
+{
+	int ret = mountpoint_redirection("print_mountpoint", CONFIG_LIB_HOMEDIR, mountpoint_handler, handler);
```

Comment:
I think you can merge mountpoint_handler & mountpoint_recursive.. is there any special reason?

Reply:
Modified as simple as possible.